### PR TITLE
#161348799 Users should be able to comment on articles

### PR DESCRIPTION
**Pull request by RachaelNantale**
#### What does this PR do?
- Comment on an article
- Get all comments on an article

#### How should this be manually tested?
POST `/api/articles/<id>/comments`
GET `/api/articles/<id>/comments`

```
{
  "comments":{
  	"comment_body": "here is another  comment for article 4"
  	
  }
}
```

Example response:
```
{
    "comments": {
        "id": 1,
        "comment_body": "Yes surely PHP is great",
        "author": {
            "id": 1,
            "username": "rachaelminani",
            "email": "rachaelminani42@gmail.com"
        },
        "created_at": "2018-11-08T12:42:19.165352Z",
        "updated_at": "2018-11-08T12:42:19.165374Z"
    }
}
```
Example of a list of comments
```
{
    "comments": [
        {
            "id": 1,
            "comment_body": "Yes surely PHP is great",
            "author": {
                "id": 1,
                "username": "rachaelminani",
                "email": "rachaelminani42@gmail.com"
            },
            
            "created_at": "2018-11-08T12:42:19.165352Z",
            "updated_at": "2018-11-08T12:42:19.165374Z"
        },
        {
            "id": 2,
            "comment_body": "Yes surely PHP is great",
            "author": {
                "id": 1,
                "username": "rachaelminani",
                "email": "rachaelminani42@gmail.com"
            },
            "created_at": "2018-11-08T12:44:02.169071Z",
            "updated_at": "2018-11-08T12:44:02.169094Z"
        },
        {
              "id": 3,
            "comment_body": "Yes surely PHP is great",
            "author": {
                "id": 1,
                "username": "rachaelminani",
                "email": "rachaelminani42@gmail.com"
            },
            "created_at": "2018-11-08T12:46:11.476372Z",
            "updated_at": "2018-11-08T12:46:11.476398Z"
        }
    ]
}

```
#### Any background context you want to provide?
- A user has to be  logged in to comment on an article
- User can comment on  both authored by me and those not authored by him/her
- You must be the author of the article in order to edit it

#### What are the relevant pivotal tracker stories?
[#161348799](https://www.pivotaltracker.com/story/show/161348799)

### CODE REVIEW

**Comment by brucemakallan** [comment]
Please fix Hound violations

File: authors/apps/articles/renderers.py [hunk 1]
```diff
@@ -26,3 +26,27 @@ def render(self, data, media_type=None, renderer_context=None):
...
+
+
+class CommentRenderer(JSONRenderer):
+    charset = 'utf-8'
+
+    def render(self, data, media_type=None, renderer_context=None):
+        '''
+        Return a dictionary with 2 entries:
+        key = "articles", value = a LIST of articles
+        key = "articlesCount", value = number of articles
```
Comment (brucemakallan) [inline authors/apps/articles/renderers.py]: Remove these comments from the Comment Renderer

File: authors/apps/articles/serializers.py [hunk 2]
```diff
@@ -14,16 +16,8 @@ def to_representation(self, data):
...
-                                self).to_representation(data)
-        if User.objects.filter(pk=int(article_details["author"])).exists():
-            user_details = User.objects.get(pk=int(article_details["author"]))
-            article_details["author"] = {
-                "id": user_details.id,
-                "username": user_details.username,
-                "email": user_details.email}
-            return article_details
-        raise APIException({"error": "User does not exist"})
+        article_details = super().to_representation(data)
```
Comment (brucemakallan) [inline authors/apps/articles/serializers.py:18]: I like the refactoring you did for `to_representation`. It prevents repetitions

Comment (RachaelNantale) [reply authors/apps/articles/serializers.py:18]: Thank you

File: authors/apps/articles/renderers.py (hunk 1 above)
Comment (RachaelNantale) [reply authors/apps/articles/renderers.py]: this is a documentation that informs the use what the Renderer does

**Comment by judeinno** [comment]
Please work on the test coverage.

**Comment by RachaelNantale** [comment]
> Please work on the test coverage.

Ok, let me work on it.

**Review by judeinno — APPROVED** [review]
Good work, working on what i requested

**Review by emugaya — CHANGES_REQUESTED** [review]
Why is the Article nested in the comment when a user tries to retrieve an articles comment?

File: authors/apps/articles/tests_articles/test_comments.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
+from rest_framework.test import APITestCase
```
Comment (emugaya) [inline authors/apps/articles/tests_articles/test_comments.py]: Use credentials in base test case.

File: authors/apps/articles/tests_articles/test_comments.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        """Test whether thread comments are returned"""
+        self.client.post(self.thread_url,
+                         self.thread_data,
+                         format='json')
+        response = self.client.get(self.thread_url,
+                                   self.thread_data,
+                                   format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_comment_not_found(self):
```
Comment (emugaya) [inline authors/apps/articles/tests_articles/test_comments.py]: Why would a user want to get a single comment?

File: authors/apps/articles/urls.py [hunk 5]
```diff
@@ -1,21 +1,30 @@
...
     # GET api/articles/id
-    # PUT api/articles/id
-    # DELETE api/articles/id
-    path('/<int:pk>', ArticleRetrieveUpdateDestroy.as_view(),
+    path('articles/<int:pk>', ArticleRetrieveUpdateDestroy.as_view(),
          name='article_by_id'),
-
     # GET api/articles/slug
-    path('/<slug:slug>', ArticleRetrieveBySlugAPIView.as_view(),
+    path('articles/<slug:slug>', ArticleRetrieveBySlugAPIView.as_view(),
```
Comment (emugaya) [inline authors/apps/articles/urls.py]: Why are you adding this?

File: authors/apps/articles/urls.py [hunk 6]
```diff
@@ -1,21 +1,30 @@
...
          name='article_by_slug'),
+    path('articles/<int:pk>/comments', CommentListCreateView.as_view(),
```
Comment (emugaya) [inline authors/apps/articles/urls.py]: Please give you methods meaningful names.

Comment (RachaelNantale) [reply authors/apps/articles/urls.py]: Let me work on this.

File: authors/apps/articles/tests_articles/test_comments.py (hunk 3 above)
Comment (RachaelNantale) [reply authors/apps/articles/tests_articles/test_comments.py]: Let me work on this

File: authors/apps/articles/urls.py (hunk 5 above)
Comment (RachaelNantale) [reply authors/apps/articles/urls.py]: It is how I had initially written the path before rebasing.

File: authors/apps/articles/tests_articles/test_comments.py (hunk 4 above)
Comment (RachaelNantale) [reply authors/apps/articles/tests_articles/test_comments.py]: This is to test when a user wants to comment on a comment(thread) and the parent comment was deleted or doesn't exist.

**Pull request merged**